File: leetcode/easy/pascals-triangle.java
// https://leetcode.com/problems/pascals-triangle/
//
// Iterative solution. May also be done recursively if you want.
//
// Complexity: runtime O(n^2), space O(1) [O(n^2) if output is considered, but
// it should be ignored for complexity calculations].

class Solution {
    public List<List<Integer>> generate(final int numRows) {
        final var result = new ArrayList<List<Integer>>(numRows);

        for (int i = 0; i < numRows; ++i) {
            final var row = new ArrayList<Integer>(numRows);

            for (int j = 0; j <= i; ++j) {
                if (j == 0 || j == i) {
                    row.add(1);
                } else {
                    row.add(result.get(i - 1).get(j - 1) +
                            result.get(i - 1).get(j));
                }
            }

            result.add(row);
        }

        return result;
    }
}

File: leetcode/easy/pascals-triangle.c
/* https://leetcode.com/problems/pascals-triangle */
/*  */
/* Complexity: runtime O(n^2), space O(1) (ignoring output). */

/**
 * Return an array of arrays of size *rm.
 * The sizes of the arrays are returned as *rns array.
 * Note: Both returned array and *rns array must be malloced, assume caller
 * calls free().
 */
int **generate(const int n, int *const restrict rm, int *restrict *const rns)
{
    int **const ret = malloc(n * sizeof *ret);
    *rm = n;
    *rns = malloc(n * sizeof **rns);
    for (int i = 0; i < n; ++i) {
        const int len = i + 1;
        (*rns)[i] = len;
        ret[i] = malloc(len * sizeof ret[i][0]);
        ret[i][len - 1] = ret[i][0] = 1;
        for (int j = 1; j < len - 1; ++j) {
            ret[i][j] = ret[i - 1][j - 1] + ret[i - 1][j];
        }
    }
    return ret;
}
